### DISCUSSION TITLE
NuGet hostRules [] needs token/password in config.js. How can password/token be secured?

### DISCUSSION THREAD

**Discussion by snowleopard24** [renovatebot/renovate#22543] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Azure Pipeline Command Line

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

password in hostRules [] for user authenticating to NuGet feed in Azure DevOps must accept pipeline variable RENOVATE_TOKEN or some other way to hide the actual token and not expose it in plainsight source controlled config.js

### What you tried so far.

token: process.env.RENOVATE_TOKEN,
hostRules: [
         {
             hostType: 'nuget',
             matchHost: 'pkgs.dev.azure.com',
             username: '', // doesn't matter for azure
             password: process.env.RENOVATE_TOKEN,
         },
     ],

### Relevant debug logs

<details><summary>Logs</summary>

```
"renovateVersion": "35.106.1"

FATAL: You must configure an Azure DevOps token, or a username and password
DEBUG: Renovate exiting
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "GkLf2OeJHDTyI-FBPd6JD",
           "msg": "You must configure an Azure DevOps token, or a username and password"
         }
       ]

```

</details>

**Comment by rarkins** [comment]
`process.env.X` is a possible way - like what you showed above. It looks like your `config.js` hasn't been read if Renovate exits and can't find your `token`, so it looks like you have an environmental problem and not a configuration problem

**Reply by snowleopard24** [reply]
hmm. when I switch the process.env.x to plain sight token it runs fine and creates dependency PRs. 
Let me see what pipeline configuration I can tweak. Thanks

**Reply by snowleopard24** [reply, marked as answer]
Okay. The issue is that config.js doesn't accept a pipeline variable if we "Change the variable type to secret". Only works when the variable is "Plain variable". 

@rarkins : do you think I need to do an additional step to make this work when the AZ pipeline variable has token as secret type?
If there is nothing then I'll close this discussion marking this as an answer. Thanks

**Reply by rarkins** [reply]
I don't think there's anything more which renovate can do. If the secret is not passed through to env then renovate can't access it

**Reply by snowleopard24** [reply]
cool, thanks @rarkins
